pennylane/qnodes/qubit.py

```diff
@@ -128,16 +128,18 @@ def _pd_analytic(self, idx, args, kwargs, **options):
             temp_var.idx = n
             op.data[p_idx] = temp_var
 
-            multiplier, shift = op.get_parameter_shift(p_idx)
+            param_shift = op.get_parameter_shift(p_idx)
 
-            # shifted parameter values
-            shift_p1 = np.r_[args, args[idx] + shift]
-            shift_p2 = np.r_[args, args[idx] - shift]
+            for multiplier, a, shift in param_shift:
 
-            # evaluate the circuit at two points with shifted parameter values
-            y2 = np.asarray(self.evaluate(shift_p1, kwargs))
-            y1 = np.asarray(self.evaluate(shift_p2, kwargs))
-            pd += (y2 - y1) * multiplier
+                # shifted parameter values
+                shift_p = np.r_[args, a * args[idx] + shift]

```

Comment:
Think I need to understand what `r_` is doing here :thinking:

Reply:
`np.r_` is a really nice way of doing efficient row-wise concatenation using slices :)

Reply:
It's essentially a slightly more efficient way of doing `np.concatenate([args, np.array([a * args[idx] + shift])])`, avoiding the need to create intermediate arrays

Reply:
Oh nice! My follow up thought was why this is a use case for doing that :thinking:
E.g.
`params = [1, 2, 3]`
`idx = 1, shift = 0.5`
This means that
`shift_p = [1, 2, 3, 2.5]`? I guess I was expecting `[1, 2.5, 3]`.

Reply:
Yeah, the gradient rules in non-tape mode are written a bit strangely 😆 Before shifting a parameter, a new 'temp' parameter is added to the circuit. It is _this_ parameter that is shifted; it is then deleted once the gradient has been computed.

The tape mode implementation is a lot 'cleaner' imho

Reply:
See the corresponding logic in `QubitQNode._pd_analytic`:

```python
def _pd_analytic(self, idx, args, kwargs, **options):
    """Partial derivative of the node using the analytic parameter shift method.
    Args:
        idx (int): flattened index of the parameter wrt. which the p.d. is computed
        args (array[float]): flattened positional arguments at which to evaluate the p.d.
        kwargs (dict[str, Any]): auxiliary arguments

    Returns:
        array[float]: partial derivative of the node
    """
    n = self.num_variables
    pd = 0.0
    # find the Operators in which the free parameter appears, use the product rule
    for op, p_idx in self.variable_deps[idx]:

        # We temporarily edit the Operator such that parameter p_idx is replaced by a new one,
        # which we can modify without affecting other Operators depending on the original.
        orig = op.data[p_idx]
        assert orig.idx == idx

        # reference to a new, temporary parameter with index n, otherwise identical with orig
        temp_var = copy.copy(orig)
        temp_var.idx = n
        op.data[p_idx] = temp_var

        param_shift = op.get_parameter_shift(p_idx)
```

Reply:
Here I mainly kept the implementation as was before.